src/components/validation/runner/telemetry.rs

```diff
@@ -57,31 +72,80 @@ impl Telemetry {
         }
     }
 
-    pub fn into_collector(
+    pub async fn into_collector(
         self,
-        task_coordinator: &TaskCoordinator<Configuring>,
+        telemetry_task_coordinator: &TaskCoordinator<Configuring>,
     ) -> TelemetryCollector {
-        let telemetry_started = task_coordinator.track_started();
-        let telemetry_completed = task_coordinator.track_completed();
-        let mut telemetry_shutdown_handle = task_coordinator.register_for_shutdown();
-
-        spawn_grpc_server(self.listen_addr, self.service, task_coordinator);
+        let telemetry_started = telemetry_task_coordinator.track_started();
+        let telemetry_completed = telemetry_task_coordinator.track_completed();
+        let mut telemetry_shutdown_handle = telemetry_task_coordinator.register_for_shutdown();
+
+        // We need a task coordinator for the gRPC server because it strictly
+        // needs to be shut down after the telemetry collector. This is because
+        // the server needs to be alive to process every last incoming event
+        // from the Vector sink that we're using to collect telemetry.
+        let grpc_task_coordinator = TaskCoordinator::new();
+        spawn_grpc_server(self.listen_addr, self.service, &grpc_task_coordinator);
+        let grpc_task_coordinator = grpc_task_coordinator.started().await;
+        debug!("All gRPC task(s) started.");
 
         let mut rx = self.rx;
         let driver_handle = tokio::spawn(async move {
             telemetry_started.mark_as_done();
 
             let mut telemetry_events = Vec::new();
-            loop {
+            'outer: loop {
                 select! {
-                    _ = telemetry_shutdown_handle.wait() => break,
+                    _ = telemetry_shutdown_handle.wait() => {
+                        // After we receive the shutdown signal, we need to wait
+                        // for two event emissions from the internal_metrics
+                        // source. This is to ensure that we've received all the
+                        // events from the components that we're testing.
+                        //
+                        // We need exactly two because the internal_metrics
+                        // source does not emit component events until after the
+                        // component_received_events_total metric has been
+                        // emitted. Thus, two events ensure that all component
+                        // events have been emitted.
+
+                        debug!("Telemetry: waiting for final internal_metrics events before shutting down.");
+
+                        let mut events_seen = 0;
+                        let current_time = chrono::Utc::now();
```

Comment:
Non-blocking nit just so I don't forget, and we can add this later: we probably want to couple this with a timeout -- say like 2-3 seconds or whatever -- so that we don't wait forever.